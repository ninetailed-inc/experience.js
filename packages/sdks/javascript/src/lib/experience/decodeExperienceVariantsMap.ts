export const decodeExperienceVariantsMap = (
  encodedExperienceVariantsMap: string
): Record<string, number> => {
  return encodedExperienceVariantsMap
    .split(',')
    .map((experienceIdWithVariant) => {
      const [experienceId, _variantIndex] = experienceIdWithVariant.split('=');

      const variantIndex = parseInt(_variantIndex);

      if (!experienceId || !variantIndex) {
        return null;
      }

      return { experienceId, variantIndex };
    })
    .filter((x): x is { experienceId: string; variantIndex: number } => !!x)
    .reduce(
      (acc, curr) => ({ ...acc, [curr.experienceId]: curr.variantIndex }),
      {}
    );
};
