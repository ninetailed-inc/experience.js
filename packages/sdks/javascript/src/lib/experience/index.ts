export * from './types';
export {
  EXPERIENCE_TRAIT_PREFIX,
  selectDistribution,
  isExperienceMatch,
  selectVariant as selectExperienceVariant,
  selectHasVariants as selectHasExperienceVariants,
  selectVariants as selectExperienceVariants,
  selectBaselineWithVariants as selectExperienceBaselineWithVariants,
  selectExperience,
  selectActiveExperiments,
} from '@ninetailed/experience.js-shared';
export { decodeExperienceVariantsMap } from './decodeExperienceVariantsMap';
export { makeExperienceSelectMiddleware } from './makeExperienceSelectMiddleware';
export * from './makeChangesModificationMiddleware';
