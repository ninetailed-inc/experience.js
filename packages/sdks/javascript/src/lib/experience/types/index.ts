export type {
  Reference,
  Baseline,
  VariantRef,
  ExperienceConfiguration,
  EntryReplacement,
  InlineVariable,
  Distribution,
  ExperienceType,
} from '@ninetailed/experience.js-shared';
