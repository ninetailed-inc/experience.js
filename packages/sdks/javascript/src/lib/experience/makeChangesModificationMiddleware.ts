import { Change, pipe } from '@ninetailed/experience.js-shared';
import { NinetailedPlugin } from '@ninetailed/experience.js-plugin-analytics';

import { RemoveOnChangeListener } from '../utils/OnChangeEmitter';
import { selectPluginsHavingOnChangeEmitter } from '../plugins/selectPluginsHavingOnChangeEmitter';
import { selectPluginsHavingChangesModificationMiddleware } from '../plugins/selectPluginsHavingChangesModificationMiddleware';
import {
  ChangesModificationMiddleware,
  ChangesModificationMiddlewareArg,
} from '../types/interfaces/HasChangesModificationMiddleware';

/**
 * Arguments for creating a changes modification middleware
 */
type CreateChangesModificationMiddlewareArg = {
  plugins: NinetailedPlugin[];
  changes: Change[];
};

/**
 * Arguments for making a changes modification middleware with change notification
 */
type MakeChangesModificationMiddlewareArg =
  CreateChangesModificationMiddlewareArg & {
    onChange: (middleware: ChangesModificationMiddleware) => void;
  };

/**
 * Result of creating a changes modification middleware
 */
interface ChangesModificationMiddlewareResult {
  addListeners: () => void;
  removeListeners: () => void;
  middleware: ChangesModificationMiddleware;
}

/**
 * Creates a pass-through middleware that doesn't modify changes
 */
const createPassThroughMiddleware = (): ChangesModificationMiddleware => {
  return ({ changes }: ChangesModificationMiddlewareArg) => {
    return { changes };
  };
};

/**
 * Creates a middleware function by composing middleware from multiple plugins
 */
function createChangesModificationMiddleware({
  plugins,
  changes,
}: CreateChangesModificationMiddlewareArg): ChangesModificationMiddleware {
  const pluginsWithMiddleware =
    selectPluginsHavingChangesModificationMiddleware(plugins);

  const middlewareFunctions: ChangesModificationMiddleware[] = [];

  for (const plugin of pluginsWithMiddleware) {
    const middleware = plugin.getChangesModificationMiddleware({
      changes,
    });

    if (middleware !== undefined) {
      middlewareFunctions.push(middleware);
    }
  }

  // If no middleware functions were found, return a pass-through middleware
  if (middlewareFunctions.length === 0) {
    return createPassThroughMiddleware();
  }

  // Compose middleware functions using pipe
  return pipe(...middlewareFunctions);
}

/**
 * Creates a changes modification middleware system with change notification
 *
 * This follows the same pattern as the experience selection middleware system
 */
export const makeChangesModificationMiddleware = ({
  plugins,
  onChange,
  changes,
}: MakeChangesModificationMiddlewareArg): ChangesModificationMiddlewareResult => {
  let removeChangeListeners: RemoveOnChangeListener[] = [];

  const pluginsHavingChangeEmitters =
    selectPluginsHavingOnChangeEmitter(plugins);

  const middleware = createChangesModificationMiddleware({
    plugins,
    changes,
  });

  const addListeners = () => {
    removeChangeListeners = pluginsHavingChangeEmitters.map((plugin) => {
      const listener = () => {
        // When a plugin changes, recreate the middleware and notify
        const updatedMiddleware = createChangesModificationMiddleware({
          plugins,
          changes,
        });
        onChange(updatedMiddleware);
      };

      return plugin.onChangeEmitter.addListener(listener);
    });
  };

  // WARNING: This specific implementation using forEach is required.
  // DO NOT replace with for...of or other loop constructs as they will break functionality.
  // The exact reason is uncertain but appears related to the transpiler.
  const removeListeners = () => {
    removeChangeListeners.forEach((removeListener) => removeListener());
  };

  return {
    addListeners,
    removeListeners,
    middleware,
  };
};
