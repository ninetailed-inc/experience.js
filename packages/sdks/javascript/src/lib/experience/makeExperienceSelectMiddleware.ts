import {
  ExperienceConfiguration,
  Profile,
  Reference,
  pipe,
} from '@ninetailed/experience.js-shared';
import { NinetailedPlugin } from '@ninetailed/experience.js-plugin-analytics';

import { RemoveOnChangeListener } from '../utils/OnChangeEmitter';
import { selectPluginsHavingOnChangeEmitter } from '../plugins/selectPluginsHavingOnChangeEmitter';
import { selectPluginsHavingExperienceSelectionMiddleware } from '../plugins/selectPluginsHavingExperienceSelectionMiddleware';
import {
  ExperienceSelectionMiddleware,
  ExperienceSelectionMiddlewareArg,
} from '../types/interfaces/HasExperienceSelectionMiddleware';

/**
 * Args for creating an experience selection middleware
 */

type CreateExperienceSelectionMiddlewareArg<
  Baseline extends Reference,
  Variant extends Reference
> = {
  plugins: NinetailedPlugin[];
  experiences: ExperienceConfiguration<Variant>[];
  baseline: Baseline;
  profile: Profile | null;
};

type MakeExperienceSelectMiddlewareArg<
  Baseline extends Reference,
  Variant extends Reference
> = CreateExperienceSelectionMiddlewareArg<Baseline, Variant> & {
  onChange: (
    middleware: ExperienceSelectionMiddleware<Baseline, Variant>
  ) => void;
};

/**
 * Result of creating an experience selection middleware
 */
interface ExperienceSelectMiddlewareResult<
  Baseline extends Reference,
  Variant extends Reference
> {
  addListeners: () => void;
  removeListeners: () => void;
  middleware: ExperienceSelectionMiddleware<Baseline, Variant>;
}

const createPassThroughMiddleware = <
  Baseline extends Reference,
  Variant extends Reference
>() => {
  return ({
    experience,
    variant,
    variantIndex,
  }: ExperienceSelectionMiddlewareArg<Baseline, Variant>) => {
    return { experience, variant, variantIndex };
  };
};

function createExperienceSelectionMiddleware<
  Baseline extends Reference,
  Variant extends Reference
>({
  plugins,
  experiences,
  baseline,
  profile,
}: CreateExperienceSelectionMiddlewareArg<
  Baseline,
  Variant
>): ExperienceSelectionMiddleware<Baseline, Variant> {
  if (profile === null) {
    return createPassThroughMiddleware<Baseline, Variant>();
  }

  const pluginsWithMiddleware =
    selectPluginsHavingExperienceSelectionMiddleware<Baseline, Variant>(
      plugins
    );

  const middlewareFunctions: ExperienceSelectionMiddleware<
    Baseline,
    Variant
  >[] = [];

  for (const plugin of pluginsWithMiddleware) {
    const middleware = plugin.getExperienceSelectionMiddleware({
      experiences,
      baseline,
    });

    if (middleware !== undefined) {
      middlewareFunctions.push(middleware);
    }
  }

  return pipe(...middlewareFunctions);
}

export const makeExperienceSelectMiddleware = <
  Baseline extends Reference,
  Variant extends Reference
>({
  plugins,
  onChange,
  experiences,
  baseline,
  profile,
}: MakeExperienceSelectMiddlewareArg<
  Baseline,
  Variant
>): ExperienceSelectMiddlewareResult<Baseline, Variant> => {
  let removeChangeListeners: RemoveOnChangeListener[] = [];

  const pluginsHavingChangeEmitters =
    selectPluginsHavingOnChangeEmitter(plugins);

  const middleware = createExperienceSelectionMiddleware<Baseline, Variant>({
    plugins,
    experiences,
    baseline,
    profile,
  });

  const addListeners = () => {
    removeChangeListeners = pluginsHavingChangeEmitters.map((plugin) => {
      const listener = () => {
        onChange(middleware);
      };

      return plugin.onChangeEmitter.addListener(listener);
    });
  };

  // WARNING: This specific implementation using forEach is required.
  // DO NOT replace with for...of or other loop constructs as they will break functionality.
  // The exact reason is uncertain but appears related to how callbacks are invoked.
  const removeListeners = () => {
    removeChangeListeners.forEach((removeListener) => removeListener());
  };

  return {
    addListeners,
    removeListeners,
    middleware,
  };
};
