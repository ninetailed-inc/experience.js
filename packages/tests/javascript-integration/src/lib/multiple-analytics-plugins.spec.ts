import { setTimeout as sleep } from 'node:timers/promises';
import { Ninetailed } from '@ninetailed/experience.js';

import { NinetailedGoogleTagmanagerPlugin } from '@ninetailed/experience.js-plugin-google-tagmanager';
import { NinetailedInsightsPlugin } from '@ninetailed/experience.js-plugin-insights';
import { NinetailedSegmentPlugin } from '@ninetailed/experience.js-plugin-segment';
import { ComponentViewEventComponentType } from '@ninetailed/experience.js-plugin-analytics';

const setup = () => {
  const gtmPlugin = new NinetailedGoogleTagmanagerPlugin();
  const insightsPlugin = new NinetailedInsightsPlugin();
  const segmentPlugin = new NinetailedSegmentPlugin();

  const ninetailed = new Ninetailed(
    { clientId: 'test' },
    { plugins: [gtmPlugin, insightsPlugin, segmentPlugin] }
  );

  return { ninetailed, gtmPlugin, insightsPlugin, segmentPlugin };
};

describe('A Ninetailed setup with multiple analytics plugins', () => {
  let ninetailed: Ninetailed;
  let insightsPlugin: NinetailedInsightsPlugin;

  beforeEach(() => {
    ({ ninetailed, insightsPlugin } = setup());
  });

  describe('Sending component views directly', () => {
    it('should send a component view event to all analytics plugins', async () => {
      await ninetailed.trackComponentView({
        experience: { id: 'experience-1', name: 'test', type: 'nt_experiment' },
        variant: { id: 'variant-1', name: 'test' },
        audience: { id: 'audience-1', name: 'test' },
        componentType: ComponentViewEventComponentType.Entry,
        element: document.createElement('div'),
        variantIndex: 1,
      });
      await sleep(5);

      // check if GTM was called e2e
      expect(window.dataLayer).toHaveLength(1);
      expect(window.dataLayer && window.dataLayer[0]).toMatchSnapshot();

      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      const eventsQueue = insightsPlugin.events;
      expect(eventsQueue).toHaveLength(1);
    });
  });
});
