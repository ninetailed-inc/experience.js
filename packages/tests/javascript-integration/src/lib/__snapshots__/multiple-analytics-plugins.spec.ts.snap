// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A Ninetailed setup with multiple analytics plugins Sending component views directly should send a component view event to all analytics plugins 1`] = `
Object {
  "event": "nt_experience",
  "ninetailed_audience": "audience-1",
  "ninetailed_component": "variant-1",
  "ninetailed_experience": "experience-1",
  "ninetailed_experience_name": "test",
  "ninetailed_variant": "variant 1",
}
`;
